src/IROperator.cpp

```diff
@@ -242,7 +360,25 @@ bool is_positive_const(const Expr &e) {
         return f->value > 0.0f;
     }
     if (const Cast *c = e.as<Cast>()) {
-        return is_positive_const(c->value);
+        Type to = c->type;
+        if (!to.is_int_or_uint() ||
+            lossless_cast(to, c->value).defined()) {
+            // Either the cast does not lose information, or it's a
+            // non-integral cast, so no overflow behavior to worry about.
+            return is_positive_const(c->value);
+        } else {
+            // May need to deal with overflow behaviors.
```

Comment:
Actually now I'm wondering if this function should handle casts at all. That seems like eager simplification, but in the contexts where is_positive_const is used I think the simplifier has already run. Perhaps this function shouldn't do anything that constant folding would do.

Reply:
e.g. I wouldn't expect is_positive_const to return true on select(false, 4, 5);

Reply:
That seems reasonable to me, in which case I can remove all cast codes (and the fuzzing test). I just want it to be correct, seeing as it produced incorrect bounds for bounds inference in one of my test cases.

Reply:
Although, one thing to note: not all calls to `is_positive_const` (and the negative version) are called on simplified expressions, notably the calls in bounds inference.

Reply:
Hrm. Then having more smarts here *does* have a positive effect. Is it feasible to insert simplification into the places that might be unsimplified in bounds? Alternatively, maybe we should just have a cheap constant-folding routine that we can call first as an alternative to full simplification. Your propagate functions just handle casts. Yet another alternative would be is_bounded_const, which would take a min and max. This would be nice because it has a natural recursive thing to do when you hit a narrowing cast: reduce the max to the maximum value of the narrower type.

Reply:
I think cheap constant folding as an alternative would be better, as I'm pretty sure no simplification is done in bounds, but constant folding would be very helpful (I often see very complicated expressions produced that simplify down with just constant folding). 

I'm not as sure as to the usefulness of is_bounded_const - where would that be used?

Reply:
Well, no simplification other than in `can_prove`, which actually might be called a lot..